src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.hints
+
+import com.intellij.codeInsight.hints.*
+import com.intellij.codeInsight.hints.presentation.InlayPresentation
+import com.intellij.codeInsight.hints.presentation.InsetPresentation
+import com.intellij.codeInsight.hints.presentation.MenuOnClickPresentation
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.ui.layout.panel
+import org.rust.ide.utils.isEnabledByCfg
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.psi.RsDotExpr
+import org.rust.lang.core.psi.RsMethodCall
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.childOfType
+import org.rust.lang.core.psi.ext.endOffset
+import org.rust.lang.core.psi.ext.parentDotExpr
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.type
+import javax.swing.JComponent
+
+class RsChainMethodTypeHintsProvider : InlayHintsProvider<RsChainMethodTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = KEY
+
+    override val name: String get() = "Chain method hints"
+
+    override val previewText: String
+        get() = """
+            let foo: Result<Option<i32>, &str> = Ok(Some(10i32));
+            foo
+              .map_err(|_| 0_u32)
+              .unwrap()
+              .and_then(|_| Some("foo"))
+              .unwrap();
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        override fun createComponent(listener: ChangeListener): JComponent {
+            return panel {}
+        }
+    }
+
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
+        object : FactoryInlayHintsCollector(editor) {
+
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, true)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (DumbService.getInstance(file.project).isDumb) return true
+                if (element !is RsElement) return true
```

Comment:
```suggestion
                if (element !is RsMethodCall) return true
```
You need only method calls so you can check the necessary condition directly